src/storage/disk_table.cc

```diff
@@ -1097,8 +1097,128 @@ void DiskTableRowIterator::SeekToFirst() {
+        for (const auto& col_name : column_key.col_name()) {
+            auto it = schema.find(col_name);
+            if (it == schema.end()) {
+                PDLOG(WARNING, "not found col_name[%s]. tid %u pid %u", col_name.c_str(), id_, pid_);
+                return false;
+            }
+            col_vec.push_back(it->second);
+        }
+
+        uint32_t inner_id = table_index_.GetAllInnerIndex()->size();
+
+        index_def = std::make_shared<IndexDef>(column_key.index_name(), table_index_.GetMaxIndexId() + 1,
+                IndexStatus::kReady, ::openmldb::type::IndexType::kTimeSerise, col_vec);
+
+
+        if (!column_key.ts_name().empty()) {
+            auto ts_iter = schema.find(column_key.ts_name());
+            index_def->SetTsColumn(std::make_shared<ColumnDef>(ts_iter->second));
+        } else {
+            index_def->SetTsColumn(std::make_shared<ColumnDef>(DEFUALT_TS_COL_NAME, DEFUALT_TS_COL_ID,
+                        ::openmldb::type::kTimestamp, true));
+        }
+
+        if (column_key.has_ttl()) {
+            index_def->SetTTL(::openmldb::storage::TTLSt(column_key.ttl()));
+        } else {
+            index_def->SetTTL(*(table_index_.GetIndex(0)->GetTTL()));
+        }
+
+        if (table_index_.AddIndex(index_def) < 0) {
+            PDLOG(WARNING, "add index failed. tid %u pid %u", id_, pid_);
+            return false;
+        }
+
+        index_def->SetInnerPos(inner_id);
+        std::vector<std::shared_ptr<IndexDef>> index_vec = {index_def};
+        auto inner_index_st = std::make_shared<InnerIndexSt>(inner_id, index_vec);
+        table_index_.AddInnerIndex(inner_index_st);
+        table_index_.SetInnerIndexPos(new_table_meta->column_key_size() - 1, inner_id);
+
+        rocksdb::ColumnFamilyOptions cfo;
+
+        if (storage_mode_ == ::openmldb::common::StorageMode::kSSD) {
+            cfo = rocksdb::ColumnFamilyOptions(ssd_option_template);
+        } else {
+            cfo = rocksdb::ColumnFamilyOptions(hdd_option_template);
+        }
+
+        cfo.comparator = &cmp_;
+        cfo.prefix_extractor.reset(new KeyTsPrefixTransform());
+
+        if (index_def->GetTTLType() == ::openmldb::storage::TTLType::kAbsoluteTime ||
+            index_def->GetTTLType() == ::openmldb::storage::TTLType::kAbsOrLat) {
+            cfo.compaction_filter_factory = std::make_shared<AbsoluteTTLFilterFactory>(inner_index_st);
+        }
+
+        rocksdb::ColumnFamilyHandle* handle;
+        cf_ds_.push_back(rocksdb::ColumnFamilyDescriptor(index_def->GetName(), cfo));
+        DEBUGLOG("add cf_name %s. tid %u pid %u", index_def->GetName().c_str(), id_, pid_);
+
+        rocksdb::Status s = db_->CreateColumnFamily(cfo, index_def->GetName(), &handle);
+        if (!s.ok()) {
+            PDLOG(WARNING, "rocksdb failed to create ColumnFamily. tid %u pid %u error %s",
+                  id_, pid_, s.ToString().c_str());
+            return false;
+        }
+
+        cf_hs_.push_back(handle);
```

Comment:
`Put` and `Iterator` will read `cf_hs_`.  It need to guarantee the thread safe

Reply:
Will there be multiple requests require to add new indexes at the same time?

Reply:
yes. `AddIndex` is a RPC call. It can be executed by multiple threads at the same time.

Can learn from the memory table:
- copy the origin `cf_hs_` and update the copied `cf_hs`
- use `std::atomic_store_explicit` to set the `cf_hs_` atomically

Reply:
Since calls from multiple threads may get the same ``inner_id``, both ``mem_table`` and ``disk_table``  may suffer from concurrent calls to ``Addindex``.
https://github.com/4paradigm/OpenMLDB/blob/f50df38b238fa96342d634ac327f7897443587c9/src/storage/mem_table.cc#L591
https://github.com/4paradigm/OpenMLDB/blob/a6e1a8d7d3f558b50ba538616c4c56b6c815b246/src/storage/disk_table.cc#L1191

Reply:
Yes. I think it is a potential thread-safe issue for memory table too. @dl239 please help verify?